### PR TITLE
RouteRefresh: refactored

### PR DESCRIPTION
**Pull request by RingerJK**
### Description
RouteRefresh: fix the broken flow 

### Changelog
```
<changelog>Fixed the Route Refresh feature which was failing on each request due to parser incompatibilities.</changelog>
```

### CODE REVIEW

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigator.kt [hunk 1]
```diff
@@ -111,7 +111,7 @@ interface MapboxNativeNavigator {
      *
      * @return True if the annotations could be updated false if not (wrong number of annotations)
      */
-    fun updateAnnotations(legAnnotationJson: String, routeIndex: Int, legIndex: Int): Boolean
+    fun updateAnnotations(legAnnotationJson: String, legIndex: Int): Boolean
```
Comment (RingerJK) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigator.kt:114]: for now the SDK supports single route only. Made it consistent with `fun setRoute`

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+@file:JvmName("DirectionsRouteEx")
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt:1]: This is probably not needed as we're accessing it from Kotlin.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 3]
```diff
@@ -57,27 +59,47 @@ internal class MapboxTripSession(
...
             }
             cancelOngoingUpdateNavigatorStatusDataJobs()
-            val setRouteJob = threadController.getMainScopeAndRootJob().scope.launch {
-                navigator.setRoute(value)?.let {
-                    routeAlerts = it.routeAlerts
+            val updateRouteJob = threadController.getMainScopeAndRootJob().scope.launch {
+                if (isSameRoute) {
+                    value?.legs()?.forEachIndexed { index, routeLeg ->
+                        routeLeg.annotation()?.toJson()?.let { annotations ->
+                            navigator.updateAnnotations(annotations, index).let { success ->
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:84]: Do we want to run this in the `Main` thread? Noting that `navigator.setRoute` switches to `NavigatorDispatcher` but `navigator.updateAnnotations` doesn't. Maybe a regression?

**Review by Guardiola31337 — COMMENTED** [review]
Left a couple of comments.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt (hunk 3 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:84]: Right, `updateAnnotations` should switch to `NavigatorDispatcher` as well.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt (hunk 2 above)
Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt:1]: using for tests

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt:1]: Java tests? :trollface:

**Review by LukasPaczos — APPROVED** [review]
LGTM

suggestion:
```<changelog>Fixed the Route Refresh feature which was failing on each request due to parser incompatibilities.</changelog>```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt (hunk 2 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt:1]: Jokes aside, I think we don't need it. Tests are Kotlin.

**Review by Guardiola31337 — APPROVED** [review]

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt (hunk 2 above)
Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt:1]: it's useful for mocking static calls smt like `mockStatic("com.package.ClassName")`

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt:1]: Are you sure? https://github.com/mapbox/mapbox-navigation-android/blob/716dcc5a9e704832a784bf99fd3736c02a0b1409/libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt#L99 `NavigatorMapperKt` doesn't have it and it works 🤔

Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/DirectionsRouteEx.kt:1]: agree we could remove it. But it doesn't make effect anyway

**Pull request merged**